cmd/vclusterctl/cmd/platform/add/vcluster.go

```diff
@@ -60,5 +67,85 @@ vcluster platform add vcluster --project my-project --all
+	localVClusters, err := cmd.localVClustersFromOptions(ctx, cmd.Log, cmd.GlobalFlags, args, cmd.AddVClusterOptions)
+	if err != nil {
+		return err
+	}
+
+	if err := cmd.validateForPreExistence(ctx, localVClusters); err != nil {
+		return err
+	}
+
+	return cli.AddVClusterHelm(ctx, cmd.Log, &cmd.AddVClusterOptions, cmd.GlobalFlags, localVClusters)
+}
+
+// validateForPreExistence checks if the vCluster already exists in the platform
+// and if it does, it checks if it is in the same project as the one specified in the command.
+func (cmd *VClusterCmd) validateForPreExistence(ctx context.Context, localVClusters []find.VCluster) error {
+	targetProject := cmd.Project
+	if targetProject == "" {
+		targetProject = "default"
+	}
+
+	byNameMap := make(map[string]bool, len(localVClusters))
+	for _, v := range localVClusters {
+		byNameMap[v.Name] = true
+	}
+
+	platformClient, err := platform.InitClientFromConfig(ctx, cmd.LoadedConfig(cmd.Log))
+	if err != nil {
+		return fmt.Errorf("new client from path: %w", err)
+	}
+
+	managementClient, err := platformClient.Management()
+	if err != nil {
+		return fmt.Errorf("create management client: %w", err)
+	}
+
+	vcInstanceList, err := managementClient.Loft().ManagementV1().VirtualClusterInstances("").List(ctx, metav1.ListOptions{})
+	if err != nil {
+		return err
+	}
+
+	for _, vClusterInstance := range vcInstanceList.Items {
+		if !byNameMap[vClusterInstance.Name] {
+			continue
+		}
+		projectFromVirtualCluster := projectutil.ProjectFromNamespace(vClusterInstance.Namespace)
+		if projectFromVirtualCluster != targetProject {
+			return fmt.Errorf("virtual cluster %q already exists in the project %q, moving virtual cluster between projects is not supported", vClusterInstance.Name, projectFromVirtualCluster)
+		}
+	}
+
+	return nil
+}
+
+func (cmd *VClusterCmd) localVClustersFromOptions(ctx context.Context, log log.Logger, globalFlags *flags.GlobalFlags, args []string, options cli.AddVClusterOptions) ([]find.VCluster, error) {
+	var vClusters []find.VCluster
+
+	if len(args) == 0 && !options.All {
+		return nil, errors.New("empty vCluster name but no --all flag set, please either set vCluster name to add one cluster or set --all flag to add all of them")
+	}
+	if options.All {
+		log.Info("looking for vCluster instances in all namespaces")
+		vClustersInNamespace, err := find.ListVClusters(ctx, globalFlags.Context, "", "", log)
+		if err != nil {
+			return nil, err
+		}
+		if len(vClustersInNamespace) == 0 {
+			log.Infof("no vCluster instances found in context %s", globalFlags.Context)
+		} else {
+			vClusters = append(vClusters, vClustersInNamespace...)
+		}
+	} else {
+		// check if vCluster exists
+		vClusterName := args[0]
+		vCluster, err := find.GetVCluster(ctx, globalFlags.Context, vClusterName, globalFlags.Namespace, log)
+		if err != nil {
+			return nil, err
+		}
+		vClusters = append(vClusters, *vCluster)
+	}
```

Comment:
nit: wondering if we could do some "early returns" here in order to eliminate some elses.